crates/throttle/src/throttle.rs

```diff
@@ -0,0 +1,295 @@
+local tat = redis.call("GET", key)
+
+if not tat then
+  tat = now
+else
+  tat = tonumber(tat)
+end
+tat = math.max(tat, now)
+
+local new_tat = tat + increment
+local allow_at = new_tat - burst_offset
+local diff = now - allow_at
+
+local throttled
+local reset_after
+local retry_after
+
+local remaining = math.floor(diff / interval) -- poor man's round
+
+if remaining < 0 then
+  throttled = 1
+  -- calculate how many tokens there actually are, since
+  -- remaining is how many there would have been if we had been able to limit
+  -- and we did not limit
+  remaining = math.floor((now - (tat - burst_offset)) / interval)
+  reset_after = math.ceil(tat - now)
+  retry_after = math.ceil(diff * -1)
+elseif remaining == 0 and increment <= 0 then
+  -- request with cost of 0
+  -- cost of 0 with remaining 0 is still limited
+  throttled = 1
+  remaining = 0
+  reset_after = math.ceil(tat - now)
+  retry_in = 0 -- retry_in is meaningless when quantity is 0
+else
+  throttled = 0
+  reset_after = math.ceil(new_tat - now)
+  retry_after = 0
+  redis.call("SET", key, new_tat, "PX", reset_after)
+end
+
+return {throttled, remaining, reset_after, retry_after, tostring(diff), tostring(interval)}
+"#,
+    )
+});
+
+fn local_throttle(
+    key: &str,
+    limit: u64,
+    period: Duration,
+    max_burst: u64,
+    quantity: Option<u64>,
+) -> Result<ThrottleResult, Error> {
+    let mut store = MEMORY
+        .get_or_init(|| Mutex::new(MemoryStore::new()))
+        .lock()
+        .unwrap();
+    let max_rate = Rate::per_period(
+        limit as i64,
+        time::Duration::try_from(period).map_err(|err| Error::Generic(format!("{err:#}")))?,
+    );
+    let mut limiter = RateLimiter::new(
+        &mut *store,
+        &RateQuota {
+            max_burst: max_burst.min(limit - 1) as i64,
+            max_rate,
+        },
+    );
+    let quantity = quantity.unwrap_or(1) as i64;
+    let (throttled, rate_limit_result) = limiter
+        .rate_limit(key, quantity)
+        .map_err(|err| Error::Generic(format!("{err:#}")))?;
+
+    // If either time had a partial component, bump it up to the next full
+    // second because otherwise a fast-paced caller could try again too
+    // early.
+    let mut retry_after = rate_limit_result.retry_after.whole_seconds();
+    if rate_limit_result.retry_after.subsec_milliseconds() > 0 {
+        retry_after += 1
+    }
+    let mut reset_after = rate_limit_result.reset_after.whole_seconds();
+    if rate_limit_result.reset_after.subsec_milliseconds() > 0 {
+        reset_after += 1
+    }
+
+    Ok(ThrottleResult {
+        throttled,
+        limit: rate_limit_result.limit as u64,
+        remaining: rate_limit_result.remaining as u64,
+        reset_after: Duration::from_secs(reset_after.max(0) as u64),
+        retry_after: if retry_after == -1 {
+            None
+        } else {
+            Some(Duration::from_secs(retry_after.max(0) as u64))
+        },
+    })
+}
+
+async fn redis_throttle(
```

Comment:
naming wise, both this and gcra_throttle are redis throttles, and both use gcra under the covers. Perhaps this should be renamed to redis_cl_throttle and gcra to something like redis_gcra_script_throttle to make the difference more obvious?